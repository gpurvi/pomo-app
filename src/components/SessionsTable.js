import React from 'react';
import moment from 'moment';
import DataRow from "./DataRow";
import '../styles/components/SessionsTable.css';

const SessionsTable = (props) => {
    let totalMillis = 0, sessionTotalCount = 0;
    const displayTime = (millis) => {
        // if under 1 hour don't show 0 hour
        if (millis < 3600000) {
            return moment(millis).utc().format('m [min]');
        }
        // if exactly whole hours don't show 0 min
        if ((millis % 3600000) === 0) {
            return moment(millis).utc().format('k [h]');
        }
        // show hour and min
        return moment(millis).utc().format('k [h] m [min]');
    };

    const displayCountTime = (count, time) => {
        return `${count} / ${displayTime(time)}`;
    };

    return (
        <table className="SessionsTable">
            <thead>
            <tr>
                <th>Name</th>
                <th>Session/Time</th>
            </tr>
            </thead>
            <tbody>
            {props.sessionData.map((session) => {
                totalMillis += session.timeInMillis;
                sessionTotalCount += session.count;
                return (
                    <DataRow
                        key={session.sessionName}
                        firstCol={session.sessionName}
                        secondCol={displayCountTime(session.count, session.timeInMillis)}
                    />
                );
            })}
            </tbody>
            <tfoot>
            <tr>
                <td>Total</td>
                <td>{displayCountTime(sessionTotalCount, totalMillis)}</td>
            </tr>
            </tfoot>
        </table>
    )
};

export default SessionsTable;