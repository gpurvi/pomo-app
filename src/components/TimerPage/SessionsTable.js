import React from 'react';
import DataRow from "./DataRow";
import normalizeDuration from './../../utils/normalizeDuration';
import {reduceSessions} from "../../utils/reduceSessions";
import '../../styles/components/table.css';
import LoadingCircle from "../common/LoadingCircle";


const SessionsTable = (props) => {
    let totalMillis = 0, sessionTotalCount = 0;
    const displayCountTime = (count, time) => {
        return `${count} / ${normalizeDuration(time)}`;
    };
    const isSessionData = props.sessionData.length > 0;
    const tableStyles = isSessionData ? 'table-striped' : undefined;
    return (
        <div className='position-relative'>
            <LoadingCircle marginTop='mt-5' height='h-75' loading={props.loading}/>
            <table className={`table text-center ${tableStyles} mt-4`}>
                <thead>
                <tr>
                    <th scope="col">Session</th>
                    <th scope="col">Count / Duration</th>
                </tr>
                </thead>
                <tbody>
                {isSessionData ? (
                    reduceSessions(props.sessionData).map((session, index) => {
                        totalMillis += session.duration;
                        sessionTotalCount += session.count;
                        return (
                            <DataRow
                                key={index}
                                firstCol={session.sessionName}
                                secondCol={displayCountTime(session.count, session.duration)}
                            />
                        );
                    })

                ) : (
                    <tr>
                        <td className='no-sessions' colSpan={2}>No sessions for this date</td>
                    </tr>
                )}
                </tbody>
                {
                    isSessionData && (
                        <tfoot className='table-tfoot'>
                        <tr>
                            <td>Total</td>
                            <td>{displayCountTime(sessionTotalCount, totalMillis)}</td>
                        </tr>
                        </tfoot>
                    )
                }
            </table>

        </div>


    )
};

export default SessionsTable;

