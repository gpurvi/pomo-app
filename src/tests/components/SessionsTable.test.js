import React from 'react';
import {shallow} from 'enzyme';
import SessionsTable from '../../components/SessionsTable';

const data = [
    {sessionName: "app", count: 5, timeInMillis: 0},
    {sessionName: "guru", count: 4, timeInMillis: 6000000},
    {sessionName: "zeta", count: 1, timeInMillis: 1500000},
    {sessionName: "null", count: 0, timeInMillis: 3600000 * 2},
];

test('renders SessionsTable correctly', () => {
    const wrapper = shallow(<SessionsTable sessionData={data}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should display correct totalCount', () => {
    const wrapper = shallow(<SessionsTable sessionData={data}/>);
    expect(wrapper.find('tfoot td').at(1).text()).toContain('10');
});

test('should display correct totalTime', () => {
    const wrapper = shallow(<SessionsTable sessionData={data}/>);
    expect(wrapper.find('tfoot td').at(1).text()).toContain('4 h 5 min');
});

test('should display correct time if less then hour', () => {
    const data = [
        {sessionName: "guru", count: 2, timeInMillis: 3500000}
    ];
    const wrapper = shallow(<SessionsTable sessionData={data}/>);
    expect(wrapper.find('tfoot td').at(1).text()).toContain('58 min');
});

test('should display correct time if whole hours', () => {
    const data = [
        {sessionName: "guru", count: 2, timeInMillis: 3600000}
    ];
    const wrapper = shallow(<SessionsTable sessionData={data}/>);
    expect(wrapper.find('tfoot td').at(1).text()).toContain('1 h');
});

