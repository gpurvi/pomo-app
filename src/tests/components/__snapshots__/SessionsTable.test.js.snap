// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SessionsTable correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SessionsTable
    sessionData={
        Array [
            Object {
              "count": 5,
              "sessionName": "app",
              "timeInMillis": 0,
            },
            Object {
              "count": 4,
              "sessionName": "guru",
              "timeInMillis": 6000000,
            },
            Object {
              "count": 1,
              "sessionName": "zeta",
              "timeInMillis": 1500000,
            },
            Object {
              "count": 0,
              "sessionName": "null",
              "timeInMillis": 7200000,
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <thead>
          <tr>
                    <th>
                              Name
                    </th>
                    <th>
                              Session/Time
                    </th>
          </tr>
</thead>,
        <tbody>
          <DataRow
                    firstCol="app"
                    secondCol="5 / 0 min"
          />
          <DataRow
                    firstCol="guru"
                    secondCol="4 / 1 h 40 min"
          />
          <DataRow
                    firstCol="zeta"
                    secondCol="1 / 25 min"
          />
          <DataRow
                    firstCol="null"
                    secondCol="0 / 2 h"
          />
</tbody>,
        <tfoot>
          <tr>
                    <td>
                              Total
                    </td>
                    <td>
                              10 / 4 h 5 min
                    </td>
          </tr>
</tfoot>,
      ],
      "className": "SessionsTable",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <th>
                        Name
            </th>
            <th>
                        Session/Time
            </th>
</tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <th>
                Name
</th>,
              <th>
                Session/Time
</th>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Name",
              },
              "ref": null,
              "rendered": "Name",
              "type": "th",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Session/Time",
              },
              "ref": null,
              "rendered": "Session/Time",
              "type": "th",
            },
          ],
          "type": "tr",
        },
        "type": "thead",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <DataRow
              firstCol="app"
              secondCol="5 / 0 min"
/>,
            <DataRow
              firstCol="guru"
              secondCol="4 / 1 h 40 min"
/>,
            <DataRow
              firstCol="zeta"
              secondCol="1 / 25 min"
/>,
            <DataRow
              firstCol="null"
              secondCol="0 / 2 h"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "app",
            "nodeType": "function",
            "props": Object {
              "firstCol": "app",
              "secondCol": "5 / 0 min",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "guru",
            "nodeType": "function",
            "props": Object {
              "firstCol": "guru",
              "secondCol": "4 / 1 h 40 min",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "zeta",
            "nodeType": "function",
            "props": Object {
              "firstCol": "zeta",
              "secondCol": "1 / 25 min",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "null",
            "nodeType": "function",
            "props": Object {
              "firstCol": "null",
              "secondCol": "0 / 2 h",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "tbody",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <td>
                        Total
            </td>
            <td>
                        10 / 4 h 5 min
            </td>
</tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <td>
                Total
</td>,
              <td>
                10 / 4 h 5 min
</td>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Total",
              },
              "ref": null,
              "rendered": "Total",
              "type": "td",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "10 / 4 h 5 min",
              },
              "ref": null,
              "rendered": "10 / 4 h 5 min",
              "type": "td",
            },
          ],
          "type": "tr",
        },
        "type": "tfoot",
      },
    ],
    "type": "table",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
                        <th>
                                    Name
                        </th>
                        <th>
                                    Session/Time
                        </th>
            </tr>
</thead>,
          <tbody>
            <DataRow
                        firstCol="app"
                        secondCol="5 / 0 min"
            />
            <DataRow
                        firstCol="guru"
                        secondCol="4 / 1 h 40 min"
            />
            <DataRow
                        firstCol="zeta"
                        secondCol="1 / 25 min"
            />
            <DataRow
                        firstCol="null"
                        secondCol="0 / 2 h"
            />
</tbody>,
          <tfoot>
            <tr>
                        <td>
                                    Total
                        </td>
                        <td>
                                    10 / 4 h 5 min
                        </td>
            </tr>
</tfoot>,
        ],
        "className": "SessionsTable",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th>
                            Name
              </th>
              <th>
                            Session/Time
              </th>
</tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <th>
                  Name
</th>,
                <th>
                  Session/Time
</th>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Name",
                },
                "ref": null,
                "rendered": "Name",
                "type": "th",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Session/Time",
                },
                "ref": null,
                "rendered": "Session/Time",
                "type": "th",
              },
            ],
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <DataRow
                firstCol="app"
                secondCol="5 / 0 min"
/>,
              <DataRow
                firstCol="guru"
                secondCol="4 / 1 h 40 min"
/>,
              <DataRow
                firstCol="zeta"
                secondCol="1 / 25 min"
/>,
              <DataRow
                firstCol="null"
                secondCol="0 / 2 h"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "app",
              "nodeType": "function",
              "props": Object {
                "firstCol": "app",
                "secondCol": "5 / 0 min",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "guru",
              "nodeType": "function",
              "props": Object {
                "firstCol": "guru",
                "secondCol": "4 / 1 h 40 min",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "zeta",
              "nodeType": "function",
              "props": Object {
                "firstCol": "zeta",
                "secondCol": "1 / 25 min",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "null",
              "nodeType": "function",
              "props": Object {
                "firstCol": "null",
                "secondCol": "0 / 2 h",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "tbody",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <td>
                            Total
              </td>
              <td>
                            10 / 4 h 5 min
              </td>
</tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <td>
                  Total
</td>,
                <td>
                  10 / 4 h 5 min
</td>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Total",
                },
                "ref": null,
                "rendered": "Total",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "10 / 4 h 5 min",
                },
                "ref": null,
                "rendered": "10 / 4 h 5 min",
                "type": "td",
              },
            ],
            "type": "tr",
          },
          "type": "tfoot",
        },
      ],
      "type": "table",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
